File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/model/gatt/client/pmd/model/PressureData.kt
package com.polar.androidcommunications.api.ble.model.gatt.client.pmd.model

import com.polar.androidcommunications.api.ble.exceptions.PmdDataParseException
import com.polar.androidcommunications.api.ble.exceptions.PmdDataParseException.Companion.toHex
import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.BlePMDClient
import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.BlePMDClient.PmdDataFieldEncoding
import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.PmdDataFrame
import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.PmdDataFrameUtils
import java.lang.Float.intBitsToFloat

internal class PressureData {

    data class PressureSample internal constructor(
        val timeStamp: ULong = 0uL,
        // Sample contains signed pressure value in bar
        val pressure: Float
    )

    val pressureSamples: MutableList<PressureSample> = mutableListOf()

    companion object {
        private const val TYPE_0_SAMPLE_SIZE_IN_BYTES = 4
        private const val TYPE_0_SAMPLE_SIZE_IN_BITS = TYPE_0_SAMPLE_SIZE_IN_BYTES * 8
        private const val TYPE_0_CHANNELS_IN_SAMPLE = 1

        fun parseDataFromDataFrame(frame: PmdDataFrame): PressureData {
            return if (frame.isCompressedFrame) {
                when (frame.frameType) {
                    PmdDataFrame.PmdDataFrameType.TYPE_0 -> dataFromCompressedType0(frame)
                    else -> throw PmdDataParseException(
                        "Pressure compressed frame type ${frame.frameType} is not supported"
                    )
                }
            } else {
                when (frame.frameType) {
                    PmdDataFrame.PmdDataFrameType.TYPE_0 -> dataFromRawType0(frame)
                    else -> throw PmdDataParseException(
                        "Pressure raw frame type ${frame.frameType} is not supported"
                    )
                }
            }
        }

        private fun dataFromCompressedType0(frame: PmdDataFrame): PressureData {
            if (frame.dataContent.isEmpty()) {
                throw PmdDataParseException(
                    "Pressure compressed TYPE_0 dataContent is empty"
                )
            }
            val pressureData = PressureData()
            val samples = BlePMDClient.parseDeltaFramesAll(frame.dataContent, TYPE_0_CHANNELS_IN_SAMPLE, TYPE_0_SAMPLE_SIZE_IN_BITS, PmdDataFieldEncoding.FLOAT_IEEE754)

            val timeStamps = PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp = frame.previousTimeStamp, frameTimeStamp = frame.timeStamp, samplesSize = samples.size, frame.sampleRate)
            for ((index, sample) in samples.withIndex()) {
                val pressure = if (frame.factor != 1.0f) intBitsToFloat(sample[0]) * frame.factor else intBitsToFloat(sample[0])
                pressureData.pressureSamples.add(PressureSample(timeStamp = timeStamps[index], pressure))
            }
            return pressureData
        }

        private fun dataFromRawType0(frame: PmdDataFrame): PressureData {
            if (frame.dataContent.isEmpty() || frame.dataContent.size % TYPE_0_SAMPLE_SIZE_IN_BYTES != 0) {
                throw PmdDataParseException(
                    "Pressure raw TYPE_0 dataContent size ${frame.dataContent.size} is not a " +
                    "non-zero multiple of expected sample size $TYPE_0_SAMPLE_SIZE_IN_BYTES. " +
                    "Data: ${frame.dataContent.toHex()}"
                )
            }
            val pressureData = PressureData()
            var offset = 0
            val step = TYPE_0_SAMPLE_SIZE_IN_BYTES

            val samplesSize = frame.dataContent.size / (step * TYPE_0_CHANNELS_IN_SAMPLE)
            val timeStamps = PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp = frame.previousTimeStamp, frameTimeStamp = frame.timeStamp, samplesSize = samplesSize, frame.sampleRate)
            var timeStampIndex = 0

            while (offset < frame.dataContent.size) {
                val pressure = PmdDataFrameUtils.parseFrameDataField(frame.dataContent.sliceArray(offset until (offset + TYPE_0_SAMPLE_SIZE_IN_BYTES)), PmdDataFieldEncoding.FLOAT_IEEE754) as Float
                offset += TYPE_0_SAMPLE_SIZE_IN_BYTES
                pressureData.pressureSamples.add(PressureSample(timeStamp = timeStamps[timeStampIndex], pressure))
                timeStampIndex++
            }
            return pressureData
        }
    }
}

File: sources/iOS/ios-communications/Sources/iOSCommunications/ble/api/model/gatt/client/pmd/model/PressureData.swift
//  Copyright © 2024 Polar. All rights reserved.

import Foundation

public class PressureData {

    struct PressureSample {
        let timeStamp: UInt64
        let pressure: Float
    }

    var samples: [PressureSample]

    init(samples: [PressureSample] = []) {
        self.samples = samples
    }

    private static let TYPE_0_SAMPLE_SIZE_IN_BYTES: UInt8 = 4
    private static let TYPE_0_SAMPLE_SIZE_IN_BITS: UInt8 = TYPE_0_SAMPLE_SIZE_IN_BYTES * 8
    private static let TYPE_0_CHANNELS_IN_SAMPLE: UInt8 = 1

    static func parseDataFromDataFrame(frame: PmdDataFrame) throws -> PressureData {
        if (frame.isCompressedFrame) {
            switch (frame.frameType) {
            case PmdDataFrameType.type_0: return try dataFromCompressedType0(frame: frame)
            default: throw PmdDataParseError(message: "Compressed FrameType: \(frame.frameType) is not supported by Pressure data parser")
            }
        } else {
            switch (frame.frameType) {
            case PmdDataFrameType.type_0: return try dataFromRawType0(frame: frame)
            default: throw PmdDataParseError(message: "Raw FrameType: \(frame.frameType) is not supported by Pressure data parser")
            }
        }
    }

    private static func dataFromCompressedType0(frame: PmdDataFrame) throws -> PressureData {
        guard !frame.dataContent.isEmpty else {
            throw PmdDataParseError(message: "Pressure compressed TYPE_0 dataContent is empty")
        }
        let pressureData = PressureData()
        let samples = Pmd.parseDeltaFramesToSamples(frame.dataContent, channels: TYPE_0_CHANNELS_IN_SAMPLE, resolution: TYPE_0_SAMPLE_SIZE_IN_BITS)
        let timeStamps = try PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp: frame.previousTimeStamp, frameTimeStamp: frame.timeStamp, samplesSize: UInt(samples.count), sampleRate: frame.sampleRate)

        var pressureSamples = [PressureSample]()

        for (index, sample) in samples.enumerated() {
            guard let rawBits = sample.first else {
                throw PmdDataParseError(message: "Pressure compressed TYPE_0 sample at index \(index) is empty")
            }
            // The compressed delta values carry the IEEE-754 float bit pattern as Int32
            let pressure = Float(bitPattern: UInt32(bitPattern: rawBits))
            pressureSamples.append(PressureSample(timeStamp: timeStamps[index], pressure: pressure))
        }

        pressureData.samples = pressureSamples
        return pressureData
    }

    private static func dataFromRawType0(frame: PmdDataFrame) throws -> PressureData {
        let step = Int(TYPE_0_SAMPLE_SIZE_IN_BYTES)
        guard !frame.dataContent.isEmpty && frame.dataContent.count % step == 0 else {
            throw PmdDataParseError(
                message: "Pressure raw TYPE_0 dataContent size \(frame.dataContent.count) is not a " +
                "non-zero multiple of expected sample size \(step)")
        }
        let pressureData = PressureData()
        let samplesSize = frame.dataContent.count / step
        var offset = 0
        var timeStampIndex = 0
        
        let timeStamps = try PmdTimeStampUtils.getTimeStamps(
            previousFrameTimeStamp: frame.previousTimeStamp,
            frameTimeStamp: frame.timeStamp,
            samplesSize: UInt(samplesSize),
            sampleRate: frame.sampleRate
        )

        while (offset < frame.dataContent.count) {
            pressureData.samples.append(PressureSample(timeStamp: timeStamps[timeStampIndex], pressure: Float(bitPattern: UInt32( frame.dataContent[offset ..< (offset + Int(TYPE_0_SAMPLE_SIZE_IN_BYTES))].withUnsafeBytes { $0.load(as: UInt32.self) }))))
            offset += Int(step)
            timeStampIndex += 1
        }
        return pressureData
    }
}
